src/test/scala/com/microsoft/hyperspace/index/HybridScanTest.scala

```diff
@@ -0,0 +1,113 @@
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.execution.datasources._
+
+import com.microsoft.hyperspace.{Hyperspace, SampleData}
+import com.microsoft.hyperspace.index.Content.Directory.FileInfo
+import com.microsoft.hyperspace.index.rules.{FilterIndexRule, JoinIndexRule}
+import com.microsoft.hyperspace.util.FileUtils
+
+class HybridScanTest extends HyperspaceSuite {
+  override val systemPath = new Path("src/test/resources/hybridScanTest")
+
+  private val sampleData = SampleData.testData
+  private val sampleParquetDataLocation = "src/test/resources/sampleparquet2"
+  private var df: DataFrame = _
+  private var hyperspace: Hyperspace = _
+  private var files: Seq[FileInfo] = _
+  private val indexConfig1 = IndexConfig("index1", Seq("clicks"), Seq("query"))
+
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    val sparkSession = spark
+    import sparkSession.implicits._
+    hyperspace = new Hyperspace(spark)
+    FileUtils.delete(new Path(sampleParquetDataLocation))
+    val dfFromSample = sampleData.toDF("Date", "RGUID", "Query", "imprs", "clicks")
+    dfFromSample.write.parquet(sampleParquetDataLocation)
+
+    df = spark.read.parquet(sampleParquetDataLocation)
+    hyperspace.createIndex(df, indexConfig1)
+
+    files = df.queryExecution.optimizedPlan.collect {
+      case LogicalRelation(
+          HadoopFsRelation(
+            location: PartitioningAwareFileIndex,
+            _,
+            dataSchema,
+            _,
+            fileFormat,
+            options),
+          _,
+          _,
+          _) =>
+        location.allFiles.map(FileInfo(_))
+    }.flatten
+    val sourcePath = new Path(files.head.name)
+    val destPath = new Path(files.head.name + ".copy")
+    sourcePath.getFileSystem(new Configuration).copyToLocalFile(sourcePath, destPath)
+  }
+
+  before {}
+
+  after {}
+
+  override def afterAll(): Unit = {
+    super.afterAll()
+    spark.sessionState.conf.unsetConf(IndexConstants.INDEX_HYBRID_SCAN_ENABLED)
+    spark.sessionState.conf.unsetConf("spark.sql.autoBroadcastJoinThreshold")
+    FileUtils.delete(new Path(sampleParquetDataLocation))
+  }
+
+  test("HybridScan filter rule test") {
+    spark.conf.set("spark.hyperspace.index.hybridscan.enabled", false)
+
+    df = spark.read.parquet(sampleParquetDataLocation)
+    val query = df.filter(df("clicks") <= 2000).select(df("query"))
+    val transformed = FilterIndexRule(query.queryExecution.optimizedPlan)
+    assert(transformed.equals(query.queryExecution.optimizedPlan), "No plan transformation.")
+
+    spark.conf.set("spark.hyperspace.index.hybridscan.enabled", true)
+    val planWithHybridScan = FilterIndexRule(query.queryExecution.optimizedPlan)
+    assert(
+      !planWithHybridScan.equals(query.queryExecution.optimizedPlan),
+      "Plan should be transformed.")
+  }
+
+  test("HybridScan join rule test") {

```

Comment:
Are we handling all the cases? For example, `attrs.size != index.indexedColumns.size` case?

Reply:
For JoinIndexRule, the condition is the requirement of a candidate index; so we could only consider the problem at applying FilterIndexRule. However, in the current PR, FilterIndex with HybridScan is handled by adding appended file paths to hdfs relation node. 

FilterIndexRule will also use "on-the-fly" shuffle & BucketUnion to support "delete" dataset or indexes on multi-relation nodes. So let me add tests for this with the next PR - hybrid scan delete support.